### DISCUSSION TITLE
모여락의 정보 뷰안에 모여락의 댓글뷰가 들어 간다는 점을 고려해서 데이터 구조 수정을 제안합니다.

### DISCUSSION THREAD

**Discussion by MARKGRAYSIRO** [DeveloperAcademy-POSTECH/MacC-GetARock#106] [아이디어 제안하기]
현재 모여락 정보 페이지의 구성은 GatheringInfoPage의 안에 View 로 GeatheringComment( 모여락의 댓글)이 들어올 예정입니다.
현재 데이터 구조상으로는 GatheringCommentInfo 안에 GatheringInfo가 들어가 있는데요, 댓글의 데이터 구조 안에 모여락 정보의 데이터가 들어가는게 아니라, 모여락 정보 데이터 안에 댓글의 데이터 구조가 들어가야 하는게 아닌가 생각이 들어서 discussion 을 열었습니다. 

설명을 잘 못하겠네요... 제가 맞는건지 확신도 없어서..ㅠ.ㅠ

지금 현재 데이터 구조입니다.
![스크린샷 2022-11-24 오전 5 27 25](https://user-images.githubusercontent.com/101700487/203640291-6a950c23-7b9c-411e-9105-9f9069ea37a6.png)
![스크린샷 2022-11-24 오전 5 27 34](https://user-images.githubusercontent.com/101700487/203640304-3b14f26f-98d7-494e-92ff-d951530d3885.png)

제안하는 데이터 구조입니다.
![스크린샷 2022-11-24 오전 5 28 21](https://user-images.githubusercontent.com/101700487/203640427-18077054-7f1a-42d3-889a-32ae88c2a91c.png)
![스크린샷 2022-11-24 오전 5 28 49](https://user-images.githubusercontent.com/101700487/203640500-78429ce2-b691-4cdd-b76f-00a50bd03c2d.png)

밤이라서 머리가 안돌아가네여... 
아무튼 이런식으로 구조를 짜게 된다면, 지금 머피와 이야기 하고있는 알맞은 댓글값을 찾기위한 탐색을 하지 않아도 될거 같아서 생각해봤습니다. 
하위 뷰가 상위뷰의 데이터를 가진것 보다는 그 반대가 더 자연스러우니까...? 우선 머리가 안돌아가서 내일 세션시간에 의논 해봐야할것 같습니다..ㅠㅠ

**Comment by EuniceNam** [comment]
https://github.com/DeveloperAcademy-POSTECH/MacC-GetARock/pull/49#issuecomment-1321095540 확인해보시면 좋을 것 같아요. 데이터 모델링이 추가된 PR에서 동시성 문제를 설명하는 짧으면서 잘 설명된 댓글입니다. 

제 이해를 덧붙여보면, 모여락이 모여락 댓글을 가질 경우, 비슷한 시기에 쓰인 두 댓글이 두 사람이 각각 다른 파일을 엑스코드 프로젝트에 추가했을 때 pbxproj 충돌 나듯이 모여락에 문제가 생길 수 있습니다. 


(아래는 잘 정리되지 못한 추가설명입니다.)
pbxproj가 충돌이 일어나는 이유( 중 하나)는 pbxproj가 하위 파일의 정보를 가지기 때문입니다. 각가의 하위 뷰컨트롤러가 "나는 이 xcode 프로젝트에 소속되어 있어!" 한다면 각자 새 뷰컨트롤러 파일을 추가해도 pbxproj는 깨지지 않겠지만, 모든 뷰컨트롤러가 그 xcode 프로젝트 소속인지는 따로 확인을 해야할 거에요. 
pbxproj는 하나고, 중복 수정이 자주 일어나지는 않아(?) 컨플릭을 해결하지만, 관리해야할 pbxproj 자체가 많고 수정이 일어나는 양이나 빈도가 각 파일보다 많다면 관리하는 것이 어려울 것 같아요. 댓글의 경우는 탐색하는 것이 저 컨플릭트 해결보다 쉬워서인 것 같아요. 

그래서 만약 탐색 비용이 동시성의 중요성보다 크다면 동시성을 해결하는 식으로 해도 될 듯해요. 유저로서 중복댓 올라가거나 잠깐 안 올라가도 클라이언트에 저장되어 있으면 신경 많이 안 써서요. 인터넷을 쓴다는 것 자체가 그런 일 아주 없기도 어렵고요. 

전통적인 관계형 데이터베이스라면 모여락 표 따로, 댓글 표 따로, 모여락id와 댓글id가 연결된 표를 따로 관리할텐데 파이어베이스 파이어스토어는 전통적인 관계형 데이터베이스가 아니고 다큐먼트 베이스여서 복잡한 쿼리가 어려워서 좀 다른가봐요. NoSQL은 관계형db에서 특정 목적을 위해 뭔가를 포기하는데 데이터 양이 많아져도 되기 위해 중복작업을 잘 막는(=은행 송금에서 일어나면 큰일나는) 무결성을 좀 포기했나봐요..
